src/storage/cache/rpc/client.rs

```diff
@@ -0,0 +1,458 @@
+        Self {
+            stream: UnsafeCell::new(stream),
+            timeout_options: timeout_options.clone(),
+            seq: AtomicU64::new(0),
+            received_keepalive_seq: AtomicU64::new(0),
+            packets_keeper: UnsafeCell::new(PacketsKeeper::new(timeout_options.clone().idle_timeout.as_secs())),
+            resp_buf: UnsafeCell::new(BytesMut::with_capacity(8 * 1024 * 1024)),
+        }
+    }
+
+    /// Recv request header from the stream
+    pub async fn recv_header(&self) -> Result<RespHeader, RpcError<String>> {
+        // Try to read to buffer
+        match self.recv_len(REQ_HEADER_SIZE).await {
+            Ok(_) => {}
+            Err(err) => {
+                debug!("Failed to receive request header: {:?}", err);
+                return Err(err);
+            }
+        }
+
+        let buffer: &mut BytesMut = unsafe {
+            transmute(self.resp_buf.get())
+        };
+        let req_header = RespHeader::decode(&buffer)?;
+        debug!("Received response header: {:?}", req_header);
+
+        Ok(req_header)
+    }
+
+    /// Receive request body from the server.
+    pub async fn recv_len(&self, len: u64) -> Result<(), RpcError<String>> {
+        let mut req_buffer: &mut BytesMut = unsafe {
+            transmute(self.resp_buf.get())
+        };
+        req_buffer.resize(len as usize, 0);
+        let reader = self.get_stream_mut();
+        match read_exact_timeout!(reader, &mut req_buffer, self.timeout_options.read_timeout).await
+        {
+            Ok(size) => {
+                debug!("Received response body len: {:?}", size);
+                return Ok(());
+            },
+            Err(err) => {
+                debug!("Failed to receive response header: {:?}", err);
+                return Err(RpcError::InternalError(err.to_string()));
+            }
+        }
+    }
+
+    /// Send a request to the server.
+    pub async fn send_data(&self, data: &[u8]) -> Result<(), RpcError<String>> {
+        let writer = self.get_stream_mut();
+        match write_all_timeout!(writer, data, self.timeout_options.write_timeout).await {
+            Ok(_) => {
+                debug!("Sent data with length: {:?}", data.len());
+                return Ok(());
+            }
+            Err(err) => {
+                debug!("Failed to send data: {:?}", err);
+                return Err(RpcError::InternalError(err.to_string()));
+            }
+        }
+    }
+
+    /// Get the next sequence number.
+    pub fn next_seq(&self) -> u64 {
+        self.seq.fetch_add(1, std::sync::atomic::Ordering::Relaxed)
+    }
+
+    pub async fn ping(&self) -> Result<(), RpcError<String>> {
+        // Send keep alive message
+        let current_seq = self.next_seq();
+        // Check keepalive is valid
+        let received_keepalive_seq = self.received_keepalive_seq.load(std::sync::atomic::Ordering::Relaxed);
+        if current_seq - received_keepalive_seq > 100 {
```

Comment:
Do not use seq here due to highly dependency to the upper level usage of ping(), use timestamp instead. Add an option for the timeout threshold.

Reply:
Thank you for your suggestion, not receiving the result of a recv for a long time can lead to an exception determination timeout.